### PR TITLE
Issue #3925 : Remove hvx_64

### LINKED ISSUE #3925: Should HVX_64 support be deprecated?
It's not clear that there is any active use of HVX_64 (as opposed to HVX_128), and any new users wouldn't start there. Should we deprecate-and-remove it? If nothing else, it would allow us to drop buildbot testing of it.

### PR DESCRIPTION
**Pull request by pranavb-ca**
- Removes hvx_64
- Adds a target feature ``hvx`` which defaults to ``Target::HVX_128``

I think we should keep ``hvx_128`` around for a little bit and before ripping it out and simply going by ``hvx``

fixes #3925

### CODE REVIEW

File: apps/hexagon_benchmarks/median3x3_generator.cpp
Comment (dsharletg) [inline apps/hexagon_benchmarks/median3x3_generator.cpp:43]: constant 128

**Review by dsharletg — CHANGES_REQUESTED** [review]
Two things that affect a ton of instances:
- Should we just make `Target::HVX_128` == `Target::HVX`, so we can simplify all instances that check these to just check `Target::HVX`?
- There are a lot of instances of `vector_size = get_target()... ? 128 : 64` that should just be `vector_size = 128` now?

**Comment by pranavb-ca** [comment]
> Two things that affect a ton of instances:
> 
> * Should we just make `Target::HVX_128` == `Target::HVX`, so we can simplify all instances that check these to just check `Target::HVX`?
> * There are a lot of instances of `vector_size = get_target()... ? 128 : 64` that should just be `vector_size = 128` now?

dang it - i thought I caught all of those instances. I grepped for HVX_64 and that's why I think I missed out some of those. Will do.

**Comment by pranavb-ca** [comment]
> Two things that affect a ton of instances:
> 
> * Should we just make `Target::HVX_128` == `Target::HVX`, so we can simplify all instances that check these to just check `Target::HVX`?

Actually that is what I have done, but in a sed frenzy I went ahead and changed ``has_feature`` to ``features_any_of``

File: src/Target.h
Comment (pranavb-ca) [inline src/Target.h:96]: @dsharlet - ``Target::HVX == Target::HVX_128``

File: apps/camera_pipe/camera_pipe_generator.cpp
Comment (dsharletg) [inline apps/camera_pipe/camera_pipe_generator.cpp:507]: Keep the HVX_128 branch, not the HVX_64 branch.

File: src/CodeGen_Hexagon.cpp
Comment (dsharletg) [inline src/CodeGen_Hexagon.cpp:50]: Remove hvx_128 from this message?

File: src/Target.cpp
Comment (dsharletg) [inline src/Target.cpp:873]: features(...)

File: src/Target.cpp
Comment (dsharletg) [inline src/Target.cpp:965]: HVX_128 -> HVX

File: src/Target.cpp
Comment (dsharletg) [inline src/Target.cpp:1053]: change to hvx?

File: test/correctness/simd_op_check_hvx.cpp
Comment (dsharletg) [inline test/correctness/simd_op_check_hvx.cpp:16]: HVX_128 -> HVX?

File: test/correctness/simd_op_check_hvx.cpp
Comment (dsharletg) [inline test/correctness/simd_op_check_hvx.cpp:661]: HVX_128 -> HVX?

File: apps/camera_pipe/camera_pipe_generator.cpp
Comment (pranavb-ca) [reply apps/camera_pipe/camera_pipe_generator.cpp:507]: I jumped the gun a little bit here. Newer cores have 4 contexts, but presently the lowest we support still has 2 contexts. I'll switch it back to `` / 2 ;``

Comment (dsharletg) [reply apps/camera_pipe/camera_pipe_generator.cpp:507]: Oh, I see. Actually, I think having more strips than cores is fine, so maybe keeping this 4 is best.

File: src/CodeGen_Hexagon.cpp
Comment (pranavb-ca) [reply src/CodeGen_Hexagon.cpp:50]: I left it in the message to indicate to users that ``hvx_128`` is still ok to use as a feature.

File: src/HexagonOffload.cpp
Comment (dsharletg) [inline src/HexagonOffload.cpp:976]: Don't need this?

File: apps/camera_pipe/camera_pipe_generator.cpp
Comment (dsharletg) [inline apps/camera_pipe/camera_pipe_generator.cpp:507]: Should this condition be negated, so its future proof?

Also, doesn't this vary between generations? I think maybe we should just unconditionally use 4. It's not a big deal to oversubscribe the contexts, maybe overhead is negligibly higher? But at the same time, we often split more than necessary anyways to get the working sets small enough to fit in the cache.

**Review by dsharletg — COMMENTED** [review]
I don't know why I can't reply to any of the existing threads on the "conversation" tab... but regarding: 

> I left it in the message to indicate to users that hvx_128 is still ok to use as a feature.

If it's going to be deprecated, wouldn't it make sense to just encourage use of `hvx`?

File: apps/camera_pipe/camera_pipe_generator.cpp
Comment (pranavb-ca) [reply apps/camera_pipe/camera_pipe_generator.cpp:507]: You are right. To keep things simple, I'll switch the code to use 4 threads always.

**Comment by pranavb-ca** [comment]
@dsharletg  - Addressed comments from the last review

**Review by dsharletg — APPROVED** [review]
Thanks for the fixes!

**Review by steven-johnson — APPROVED** [review]

**Comment by steven-johnson** [comment]
LGTM -- pushed a partial change to the buildbot to avoid the hvx64 build errors, then synced this to master. Let's let it go green before landing.

**Comment by dsharletg** [comment]
This is failing on some of the buildbots:

```
make: *** No rule to make target 'bin/build/initmod.hvx_64_ll.o', needed by 'bin/libHalide.so'.  Stop.
make: *** Waiting for unfinished jobs....
```

Looking at the PR, I don't see any changes to either Makefile or CMakeLists.txt, both should probably change.

**Comment by pranavb-ca** [comment]
https://github.com/halide/build_bot/pull/108

**Comment by dsharletg** [comment]
I think that's a different thing. Yes, the hvx_64 tests will fail to run. But here, libHalide.so is failing to link at all. I think because these two lines need to be removed:

https://github.com/halide/Halide/blob/master/Makefile#L829
https://github.com/halide/Halide/blob/master/src/runtime/CMakeLists.txt#L105

**Comment by pranavb-ca** [comment]
> I think that's a different thing. Yes, the hvx_64 tests will fail to run. But here, libHalide.so is failing to link at all. I think because these two lines need to be removed:
> 
> https://github.com/halide/Halide/blob/master/Makefile#L829
> https://github.com/halide/Halide/blob/master/src/runtime/CMakeLists.txt#L105

Yes, you are right. I was simply reminded of my PR from yesterday because of your comment and so pasted that here. I had taken care of CMakeLists.txt in a previous commit. Just pushed one more change to remove it from Halide/Makefile

**Comment by alexreinking** [comment]
> I think we should keep `hvx_128` around for a little bit and before ripping it out and simply going by `hvx`

If we hold off merging this until v11.0.0 is released, then we don't need to wait at all.

**Comment by dsharletg** [comment]
This has passed all the build bots, @alexreinking is there really a reason to wait to merge this? It isn't a breaking change (unless you are using hvx_64 of course). hvx_128 remains as a useful flag, it's just deprecated (and it is only silently deprecated at this point). I think the steps here are:

1. Remove `hvx_64` (this PR)
2. Deprecate `hvx_128` by warning people to use `hvx` instead.
3. Remove `hvx_128` and only support `hvx`.

I think we should do 1 now that this PR is ready to merge, and avoid generating inevitable merge conflicts for this PR to manage later.

2 and 3 aren't planned for any particular time yet. 2 and 3 should probably be in separate releases of course.

**Comment by alexreinking** [comment]
It's totally fine to put this through a deprecation plan. I thought this PR might not be ready for the v11.0 release, but it looks like I misread the room :)

**Comment by dsharletg** [comment]
OK, I think I'll let @pranavb-ca decide when to merge it. I was also mistaken, some of the build bots are still running.

**Comment by steven-johnson** [comment]
I agree with @dsharlet -- happy to go ahead and land before branching the 11 release. Buildbots are clean, there's just one merge conflict to resolve.

**Comment by steven-johnson** [comment]
UPDATE: the merge conflict was due to my clang-tidy changes, so I went ahead and pushed the fix. Should be good to merge anytime.

**Comment by steven-johnson** [comment]
Green, OK to land

**Pull request merged**
